[influxdb] Add option for using metadata value as measurement name

DerOetzi opened the pull request:
Add option for using the metadata value as measurement name. Part of solution for #9556 

Signed-off-by: Johannes Ott <info@johannes-ott.net>

lujop reviewed (commented):
Thanks for your contribution @DerOetzi 
But I see some problems with this because it will break the query part of the addon:

- A complete solution should add also itemName as a tag when this option is used and make query part intelligent to use that logic.
- Also I would like to include test covering that new functionality.
- I added also some minor notes in your code

lujop reviewed (changes requested):
Changes explained in previous comment

DerOetzi commented on bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -81,6 +79,14 @@
 			<default>false</default>
 		</parameter>
 
+		<parameter name="metaMeasurementName" type="boolean" required="true" groupName="misc">
```
Renamed

DerOetzi commented:
> Thanks for your contribution @DerOetzi
> But I see some problems with this because it will break the query part of the addon:
> 
> * A complete solution should add also itemName as a tag when this option is used and make query part intelligent to use that logic.

The itemname is already always included as tag, if I see this correct. But your right I missed the query part. Please review my changes on this.

> * Also I would like to include test covering that new functionality.

I have added some tests. Please give me feedback, whether they are enough like this.

> * I added also some minor notes in your code

Renamed the things you commented

Hilbrand commented on bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreator.java:
```diff
@@ -64,6 +64,20 @@ public ItemToStorePointCreator(InfluxDBConfiguration configuration, @Nullable Me
     private String calculateMeasurementName(Item item, @Nullable String storeAlias) {
         String name = storeAlias != null && !storeAlias.isBlank() ? storeAlias : item.getName();
 
+        if (configuration.isUseMetadataMeasurementName()) {
+            final MetadataRegistry currentMetadataRegistry = metadataRegistry;
+            if (currentMetadataRegistry != null) {
+                MetadataKey key = new MetadataKey(InfluxDBPersistenceService.SERVICE_NAME, item.getName());
+                Metadata metadata = currentMetadataRegistry.get(key);
+                if (metadata != null) {
+                    String metaName = metadata.getValue();
+                    if (!metaName.isBlank()) {
+                        name = metaName;
+                    }
+                }
+            }
+        }
```
This code to search te MetadataRegistry is 3 times the same. Can you put it in a method that used at all these 3 locations?

DerOetzi replied:
Refactored to a new utils class

DerOetzi commented on bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -81,6 +79,14 @@
 
+		<parameter name="metaMeasurementName" type="boolean" required="true" groupName="misc">
+			<label>Use metadata name</label>
+			<description>Should the measurement name set by metadata influxdb value if exists? If no metadata is set, item name
```
@lujop Can you have a look on the new description?

DerOetzi replied:
Isn't relevant anymore!

Hilbrand reviewed (approved):
LGTM. @lujop Can you have a second look at this pull request and approve if ok.

lujop commented on bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -81,6 +81,14 @@
 			<default>false</default>
 		</parameter>
 
+		<parameter name="useMetadataMeasurementName" type="boolean" required="true" groupName="misc">
```
Add that parameter also to the documentation in the README. 
And a reference where it says ` with names equal to the...`
Also, it's important to document in the README the name of the metadata key the user must use.

DerOetzi replied:
Will add this and all other Metadata configuration to README as a complete new section.

> And a reference where it says with names equal to the...

I don't understand what you want me to do? Can you please explain?

lujop replied:
In line 9 of the README:
Change:

    - The states of an item are persisted in _measurements_ points with names equal to the name of the item, or the alias, if one is provided. In both variants, a _tag_ named "item" is added, containing the item name.

for
  - The states of an item are persisted in _measurements_ points with names equal to the name of the item, its alias, or from some metadata depending on the configuration. In all variants, a _tag_ named "item" is added, containing the item name.

DerOetzi replied:
Changed

lujop reviewed (changes requested):
Thank you for your modifications, I haven't had time to test but it seems ok.
Only two minor things.

DerOetzi commented on bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBMetadataUtils.java:
```diff
@@ -0,0 +1,55 @@
+            return name;
+        }
+
+        if (configuration.isUseMetadataMeasurementName()) {
```
Okay will change this!

DerOetzi replied:
While writing documentation about usage of metadata for additional information, I wonder now whether this configuration option is needed at all. User will configure it's item like this:

```
Number testItem1 {... influxdb= "" [tag1=foo, tag2= bar]}
or
Number testItem2 {... influxdb= "measurement_name" [tag1=foo, tag2=bar]}
```
Thinking again about this, it could be handled without this additional configuration? What do you think @lujop and @Hilbrand?

DerOetzi commented:
I did some refactoring of my code as mentioned above there is no configuration flag for service anymore, but it just checks whether there is a metadata value or not. 

But while doing the refactoring I think I have found a bug inside the existing code, of ItemToStorePointCreator which wasn't noticed because of a missing test case. But I'm not sure and need your experience with the code @lujop and @Hilbrand.

At the `InfluxFilterCriteriaQueryCreatorImplTest.java`  there is a testcase `testEscapeSimpleItem` which test an item with a dot in its name. This one fails now with an IllegalArgumentException when trying to generate MetadataKey from item name.

I have added now such testcase to the ItemToStorePointCreator file as well, which fails although with the old code, without my changes because it tries to access MetadataKey in addPointTags as well.

My question is it a valid item name with a dot, then the question is how to access metadata with such items, or is this testcase invalid in the FilterQuery test?

DerOetzi commented:
I did some further research from what I see so far an item name cannot contain a dot because this will always fail here:

https://github.com/openhab/openhab-core/blob/029280683087d07d3d52bc1b125e73c17a33917a/bundles/org.openhab.core/src/main/java/org/openhab/core/common/AbstractUID.java#L96

So I think the testcase is not valid and I would remove it?

DerOetzi commented on bundles/org.openhab.persistence.influxdb/src/test/java/org/openhab/persistence/influxdb/internal/ItemToStorePointCreatorTest.java:
```diff
@@ -80,6 +80,14 @@ public void convertBasicItem(Number number) {
         return Stream.of(5, 5.5, 5L);
     }
 
+    @Test
```
As described in my comment above this test will fail although with the old code, because I think a item name with a dot inside is not valid

https://github.com/openhab/openhab-core/blob/029280683087d07d3d52bc1b125e73c17a33917a/bundles/org.openhab.core/src/main/java/org/openhab/core/common/AbstractUID.java#L96

But in filter criteria Test there is an existing testcase which fails now as well.

DerOetzi commented:
@lujop and @Hilbrand removed the testcase for item name with dot

lujop commented on bundles/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/influx1/Influx1FilterCriteriaQueryCreatorImpl.java:
```diff
@@ -61,7 +61,7 @@ public String createQuery(FilterCriteria criteria, String retentionPolicy) {
         Where where = select.where();
 
-        if (configuration.isUseMetadataMeasurementName() && hasCriteriaName) {
+        if (!(itemName == null || tableName.equals(itemName))) {
```
```suggestion
        if (itemName != null && !tableName.equals(itemName))) {
```

lujop reviewed (approved):
First of all, I apologize for not being able to review the addon before.
Today I've been able to build and test it for InfluxDB 2 and worked correctly.

The code seems good to me also. 
I only made a little commit with some minor documentation changes and suggested a change in a condition expanding the negation to make  reading the condition clearer.

Thanks for the PR

lujop commented on bundles/org.openhab.persistence.influxdb/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -81,6 +81,14 @@
 			<default>false</default>
 		</parameter>
 
+		<parameter name="useMetadataMeasurementName" type="boolean" required="true" groupName="misc">
+			<label>Use metadata name</label>
+			<description>If true uses metadata from item influxdb metadata value as name for measurement instead of item name. If
+				metadata value is not set item name is used.
+			</description>
+			<default>false</default>
+		</parameter>
+
```
```suggestion
```
In the review I forgot to say that his must be removed as it isn't any longer used.

lujop reviewed (changes requested):
Just remove the unused parameter from config.

DerOetzi commented:
> First of all, I apologize for not being able to review the addon before.
> Today I've been able to build and test it for InfluxDB 2 and worked correctly.
> 
> The code seems good to me also.
> I only made a little commit with some minor documentation changes and suggested a change in a condition expanding the negation to make reading the condition clearer. Please accept the commit suggestion.
> 
> Good work, and thank you very much for the PR.

Thanks for your constructive review comments. Think we found a good solution together now.

DerOetzi commented:
Please don't merge yet I think I have found another problem on restoring item states

DerOetzi commented:
> Please don't merge yet I think I have found another problem on restoring item states
 
@lujop can you please recheck my last changes to V1 Query parts. As I noticed there was a bug because tem-tags where not accquired.

lujop commented:
> > Please don't merge yet I think I have found another problem on restoring item states
> 
> @lujop can you please recheck my last changes to V1 Query parts. As I noticed there was a bug because tem-tags where not accquired.

Sure, let me find some time and I will install InfluxDB1 and do some more detailed testing

DerOetzi commented:
> > > Please don't merge yet I think I have found another problem on restoring item states
> > 
> > 
> > @lujop can you please recheck my last changes to V1 Query parts. As I noticed there was a bug because tem-tags where not accquired.
> 
> Sure, let me find some time and I will install InfluxDB1 and do some more detailed testing

I have installed a v1.8 docker container and did some more testing myself. For me everything seems to work fine now, but I would be very glad, if find the time to check it as well @lujop

lujop reviewed (approved):
I've done some testing with InfluxDB1 and it seems to be working correctly.
Thanks for the fixs.

lujop commented:
@Hilbrand It seems ok to be integrated for me.

DerOetzi commented:
@Hilbrand and @lujop is there any problem with this PR? Why isn't it merged yet?

kaikreuzer commented:
As @Hilbrand already approved on Feb 3, I'd say we can merge. 👍
